application-task-default/src/main/java/com/xwiki/task/internal/notifications/TaskChangedEventNotificationListener.java

```diff
@@ -0,0 +1,218 @@
+     */
+    private static final List<String> WATCHED_FIELDS =
+        Arrays.asList(Task.ASSIGNEE, Task.DUE_DATE, Task.PROJECT, Task.STATUS, Task.SEVERITY);
+
+    @Inject
+    private ComponentManager componentManager;
+
+    private ObservationManager observationManager;
+
+    /**
+     * Initialize the listener.
+     */
+    public TaskChangedEventNotificationListener()
+    {
+        super(TaskChangedEventNotificationListener.class.getName(),
+            Arrays.asList(new XObjectUpdatedEvent(CLASS_MATCHER), new XObjectAddedEvent(CLASS_MATCHER)));
+    }
+
+    /**
+     * Helper to get the value of the properties of an XObject.
+     * 
+     * @param obj the base object
+     * @param propertyName the property to retrieve
+     * @return the value of the desired property, or null if the property doesn't exist or has unsupported type.
+     */
+    private static Object getPropertyValue(BaseObject obj, String propertyName)
+    {
+        if (null == obj) {
+            return null;
+        }
+        // Method getValue() always returns null from the Property interface implementation, so this if chain is needed.
+        PropertyInterface property = obj.safeget(propertyName);
+        if (property instanceof StringProperty) {
+            return ((StringProperty) property).getValue();
+        } else if (property instanceof LargeStringProperty) {
+            return ((LargeStringProperty) property).getValue();
+        } else if (property instanceof DateProperty) {
+            return (Date) ((DateProperty) property).getValue();
+        } else {
+            return null;
+        }
+    }
+
+    /**
+     * Get a {@link TaskChangedEvent} reflecting the changes made to a certain property, if a change happened.
+     * 
+     * @param currentObject the current version of the object
+     * @param previousObject the previous version of the object
+     * @param propertyName the property to check for changes
+     * @param baseEvent a skeleton event used as a template when calling multiple times for the same object
+     * @return the event describing the changes done to the specified field, null when no change occured
+     */
+    private TaskChangedEvent getFieldChangedEvent(BaseObject currentObject, BaseObject previousObject,
+        String propertyName, TaskChangedEvent baseEvent)
+    {
+        Object currentValue = getPropertyValue(currentObject, propertyName);
+        Object previousValue = getPropertyValue(previousObject, propertyName);
+
+        if (currentValue == null) {
+            return null;
+        }
+
+        boolean valuesAreEqual = currentValue.equals(previousValue);
+
+        String localizationSuffix = propertyName;
+
+        if (currentValue instanceof Date) {
+            long currentDate = ((Date) currentValue).getTime();
+            long previousDate = ((Date) previousValue).getTime();
+
+            valuesAreEqual = (currentDate == previousDate);
+            if (currentDate > previousDate) {
```

Comment:
Some things can only be done in velocity, so they will probably have to remain there (like changing the message if the current user is the one being assigned).
A case could be made for keeping the date conversion in velocity as well, so that changing the `datemacro` date format in the admin settings will not result in inconsistent dates depending on when the event was generated. Currently the date is passed as UNIX time, which can be (re)formatted at rendering time without this issue.

Reply:
> like changing the message if the current user is the one being assigned

You can get this type of information in here as well, using `context.getUserReference()`. Is something else that prevents you from moving the code here?

I was looking for a solution to do all in here (or all in velocity if that's better) because right now it seems a little out of scope having this date field value handling in it (I would move it in a separated method anyway)

Reply:
The translation key computed in java can't be changed afterwards without modifying the event. I'm hesitant to edit it in java because if the user is set here, then all users subscribed to the page would see the 'You were assigned' text, even if it was another user.

I'll move everything to velocity so it's all together.